ui/StatusQ/src/submodelproxymodel.cpp

```diff
@@ -126,28 +171,125 @@ void SubmodelProxyModel::setSubmodelRoleName(const QString& sumodelRoleName)
     initializeIfReady();
 }
 
+void SubmodelProxyModel::onCustomRoleChanged(QObject* source, int role)
+{
+    if (!m_dataChangedQueued) {
+        m_dataChangedQueued = true;
+        QMetaObject::invokeMethod(this, "emitAllDataChanged", Qt::QueuedConnection);
+    }
+}
+
+void SubmodelProxyModel::emitAllDataChanged()
+{
+    m_dataChangedQueued = false;
+    auto count = rowCount();
+
+    if (count == 0)
+        return;
+
+    QVector<int> roles(m_additionalRolesMap.cbegin(),
+                       m_additionalRolesMap.cend());
+
+    emit this->dataChanged(index(0, 0), index(count - 1, 0), roles);
+}
+
 void SubmodelProxyModel::initializeIfReady()
 {
     if (!m_submodelRoleName.isEmpty() && sourceModel()
-            && !roleNames().empty())
+            && !sourceModel()->roleNames().empty() && m_delegateModel)
         initialize();
 }
 
 void SubmodelProxyModel::initialize()
 {
-    auto roles = roleNames();
-    auto keys = roles.keys(m_submodelRoleName.toUtf8());
-    auto keysCount = keys.size();
-
-    if (keysCount == 1) {
-        m_initialized = true;
-        m_submodelRole = keys.first();
-    } else if (keysCount == 0){
+    auto roles = sourceModel()->roleNames();
+    auto submodelKeys = roles.keys(m_submodelRoleName.toUtf8());
+    auto submodelKeysCount = submodelKeys.size();
+
+    if (submodelKeysCount == 1) {
+        m_submodelRole = submodelKeys.first();
+    } else if (submodelKeysCount == 0){
         qWarning() << "Submodel role not found!";
+        return;
     } else {
         qWarning() << "Malformed source model - multiple roles found for given "
                       "submodel role name!";
+        return;
+    }
+
+    auto creationContext = m_delegateModel->creationContext();
+    auto parentContext = creationContext
+            ? creationContext : m_delegateModel->engine()->rootContext();
+
+    QIdentityProxyModel emptyModel;
+
+    auto context = std::make_unique<QQmlContext>(parentContext);
+
+    // The delegate object is created in order to inspect properties. It may
+    // be not properly initialized because of e.g. lack of context properties
+    // containing submodel. To avoid warnings, they are muted by setting empty
+    // message handler temporarily.
+    QtMessageHandler originalHandler = qInstallMessageHandler(
+                emptyMessageHandler);
+    std::unique_ptr<QObject> instance(m_delegateModel->create(context.get()));
+    qInstallMessageHandler(originalHandler);
+
+    const QMetaObject* meta = instance->metaObject();
+
+    QStringList additionalRoles;
+
+    for (auto i = meta->propertyOffset(); i < meta->propertyCount(); ++i) {
+        const QString propertyName = meta->property(i).name();
```

Comment:
AS QMetaProperty::name returns const char*, would it make sense to use QLatin1String instead of QString?

Reply:
And that's why it's always better to use `const auto`, the compiler is smarter than us :)